libs/mock-block-dock/src/datastore/traverse.ts

```diff
@@ -1,284 +1,543 @@
+             */
+            entityId: mustBeDefined(leftEntityRevisions[0]).metadata.recordId
+              .entityId,
+          },
+        };
+
+        patchedAddKnowledgeGraphEdge(
+          traversalSubgraph,
+          sourceEntityId,
+          hasLeftEntityEdge,
+        );
+
+        return leftEntityRevisions.map((entity) => ({
+          kind: "entity",
+          inner: entity,
+        }));
+      }
+
+      return [];
+    }
+    case "HAS_RIGHT_ENTITY": {
+      if (!isEntityVertex(source)) {
+        return [];
+      }
+
+      const { inner: sourceEntity } = source;
+
+      const sourceEntityId = sourceEntity.metadata.recordId.entityId;
+
+      if (reversed) {
+        // get incoming links for entity
+        const incomingLinks = getIncomingLinksForEntity(
+          datastore,
+          sourceEntityId,
+          interval,
+        );
+        const mappedRevisions = mapElementsIntoRevisions(incomingLinks);
+
+        for (const incomingLinkEntityId of Object.keys(mappedRevisions)) {
+          const incomingLinkEdge: PatchedIncomingLinkEdge = {
+            kind: "HAS_RIGHT_ENTITY",
+            reversed: true,
+            rightEndpoint: {
+              entityId: incomingLinkEntityId,
+            },
+          };
+
+          patchedAddKnowledgeGraphEdge(
+            traversalSubgraph,
+            sourceEntityId,
+            incomingLinkEdge,
+          );
+        }
+
+        return Object.values(mappedRevisions)
+          .flat()
+          .map((entity) => ({
+            kind: "entity",
+            inner: entity,
+          }));
+      } else if (
+        sourceEntity.linkData?.leftEntityId !== undefined &&
+        sourceEntity.linkData?.rightEntityId !== undefined
+      ) {
+        // get right entity for link entity
+        const rightEntityRevisions = getRightEntityForLinkEntity(
+          datastore,
+          sourceEntityId,
+          interval,
+        );
+        const hasRightEntityEdge: PatchedHasRightEntityEdge = {
+          kind: "HAS_RIGHT_ENTITY",
+          reversed: false,
+          rightEndpoint: {
+            /**
+             * @todo - This assumes that once a link entity's endpoints cannot change. This is based on the assumption
+             *   that once entity is a link, it's always a link, however we don't enforce this currently.
+             */
+            entityId: mustBeDefined(rightEntityRevisions[0]).metadata.recordId
+              .entityId,
+          },
+        };
+
+        patchedAddKnowledgeGraphEdge(
+          traversalSubgraph,
+          sourceEntityId,
+          hasRightEntityEdge,
+        );
+
+        return rightEntityRevisions.map((entity) => ({
+          kind: "entity",
+          inner: entity,
+        }));
+      }
+
+      return [];
+    }
+    default: {
+      throw new Error(`Currently unsupported traversal edge kind: ${edgeKind}`);
```

Comment:
Is it intended, that we cannot traverse from `entity` to `entityType`? Will this be added in a follow-up?